homeassistant/components/imap/coordinator.py

```diff
@@ -390,15 +417,23 @@ async def _async_wait_push_loop(self) -> None:
                 await asyncio.sleep(BACKOFF_TIME)
                 continue
             else:
+                self.auth_errors = 0
                 self.async_set_updated_data(number_of_messages)
+                if self.config_entry.state == ConfigEntryState.SETUP_ERROR:
+                    self.config_entry.async_set_state(
+                        self.hass,
+                        ConfigEntryState.LOADED,
+                        None,
+                    )
             try:
                 idle: asyncio.Future = await self.imap_client.idle_start()
                 await self.imap_client.wait_server_push()
                 self.imap_client.idle_done()
                 async with async_timeout.timeout(10):
                     await idle
 
-            except (AioImapException, asyncio.TimeoutError):
+            # From python 3.11 asyncio.TimeoutError is an alias of TimeoutError
+            except (AioImapException, asyncio.TimeoutError, asyncio.CancelledError):
```

Comment:
We shouldn't trap `CancelledError` because it can block cancellation.

If you need to ensure cleanup happens, let the cancel raise and cleanup in a `finally`